### ISSUE TITLE
speedup install stages on travis

### ISSUE THREAD

**Issue by efiop** [iterative/dvc#2361]
E.g. https://travis-ci.com/iterative/dvc/jobs/221852201
<img width="1050" alt="Screen Shot 2019-08-02 at 8 38 04 PM" src="https://user-images.githubusercontent.com/5367102/62388260-7d4c7d80-b565-11e9-8697-fdc4e2f178b0.png">
https://travis-ci.com/iterative/dvc/jobs/221852207
<img width="1047" alt="Screen Shot 2019-08-02 at 8 38 49 PM" src="https://user-images.githubusercontent.com/5367102/62388275-8fc6b700-b565-11e9-9c01-72d0d5f98293.png">

**Comment by mroutis** [comment]
Maybe we can use the addon on `macos` to update brew itself and install pyenv
https://docs.travis-ci.com/user/installing-dependencies/#installing-packages-on-macos

**Comment by mroutis** [comment]
Not sure if `.` includes a `.venv` or just the source code, if so, testing for `.` is slower than doing `black dvc tests --check` and `flake8 dvc tests`
```
black ./ --check
flake8 ./
```

**Comment by mroutis** [comment]
This might not speed up anything, but the script will be less complex if adding keys to known hosts are described on the yaml:
https://docs.travis-ci.com/user/ssh-known-hosts/

**Comment by mroutis** [comment]
```
if [ "$TRAVIS_OS_NAME" == "linux" ] && [ "$TRAVIS_PULL_REQUEST" == "false" ]
then
  bash "$scriptdir/install_azurite.sh"
  bash "$scriptdir/install_oss.sh"
  bash "$scriptdir/install_hadoop.sh"
fi
```

Looks like we are only testing cloud providers under linux, making cached Docker Images an option :thinking:

**Comment by efiop** [comment]
Need to take a look at windows builds, they are our bottleneck.

**Comment by efiop** [comment]
as @mroutis pointed out, the cause for slow windows installs might be that some packages don't have windows wheels.

**Comment by efiop** [comment]
Alright, so in https://github.com/iterative/dvc/pull/2398 I've tried:

1) caching pip(already happening in master). This helps and works already. The problem is that it is not the download that takes time(or even wheel building, those are also cached already), but the installing itself.

2) caching site-packages or `virtualenv --relocatable`. This didn't help because we need to reinstall some packages each time(e.g. azurite), and cacher was always detecting changes and creating new cache. Though, even unpacking it was taking around 4-5 minutes anyway due to the number of packages that we need all all their unpacked files, which defeats the purpose.

3) parallelizing pip install with xargs or pipenv. This didn't help because we need to basically install everything through pip, because basic python images on windows don't have anything installed (pip freeze literally shows 0 packages), so every package was trying to install the same list of basic packages and so the performance benefit was marginal, but script readability and debugging was decreasing very quickly.

So in conclusion, until travis guys bring python in the base image, we will have to deal with what we have right now, which is under 30 minutes for a full build, bottlenecked by windows builds https://travis-ci.community/t/python-support-on-windows/241/17
